[feature/keyboard-commands] Keyboard Commands for all actions on iPad

hosy opened the pull request:
## Description
All available actions should be handled by keyboard commands. 
Also added:
- complete navigation, selection
- copy and paste files/items with keyboard command into the app and to/from other apps

## Related Issue
#282 

## Motivation and Context
Completely control the app with a hardware keyboard

## How Has This Been Tested?
At first connect a hardware keyboard to your iPad (bluetooth or wired).
To show available keyboard commands, long press the `Command` key. This will show you, which actions are currently possible (depends on selected items).

## Screenshots (if appropriate):
![IMG_139895F2FD9C-1](https://user-images.githubusercontent.com/736109/64124320-fcbcaf00-cda6-11e9-8321-de01a29572a6.jpeg)

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING**](https://github.com/owncloud/ios-app/blob/master/CONTRIBUTING.md) document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### QA

Test plan:

https://github.com/owncloud/QA/blob/master/Mobile/iOS-app/Version%201.2.0/Keyboard.md

BUGS & IMPROVEMENTS

- [X] (1) Not posible to submit the target folder in copy/move https://github.com/owncloud/ios-app/pull/495#issuecomment-545471975 [FIXED]
- [X] (2) No way to select expiration date in public links https://github.com/owncloud/ios-app/pull/495#issuecomment-545476782 [WONT FIX]
- [x] (3) Searching only posible if no items selected https://github.com/owncloud/ios-app/pull/495#issuecomment-545479159 [FIXED]
- [X] (4) No way to create folder/upload content in root folder https://github.com/owncloud/ios-app/pull/495#issuecomment-545480407 [FIXED]
- [x] (5) Making av. offline. makes everything in the list https://github.com/owncloud/ios-app/pull/495#issuecomment-545485797 [FIXED]
- [x] (6) Missing options https://github.com/owncloud/ios-app/pull/495#issuecomment-545487365 [FIXED]
- [X] (7) Authentication issues https://github.com/owncloud/ios-app/pull/495#issuecomment-546845722 [FIXED]
- [X] (8) Cancel manage account view https://github.com/owncloud/ios-app/pull/495#issuecomment-546846783 [FIXED]
- [X] (9) Delete unused copies in Settings https://github.com/owncloud/ios-app/pull/495#issuecomment-546848237 [FIXED]
- [x] (10) Crash in file picker when the end of the list is reached https://github.com/owncloud/ios-app/pull/495#issuecomment-547817544 [FIXED]
- [x] (11) Crash in the welcome view with the down arrow https://github.com/owncloud/ios-app/pull/495#issuecomment-547982345 [FIXED]
- [x] (12) Add account command lost https://github.com/owncloud/ios-app/pull/495#issuecomment-548339638 [NOT REPRODUCIBLE]
- [x] (13) Cancel delete does not work https://github.com/owncloud/ios-app/pull/495#issuecomment-548345187  [FIXED]
- [x] (14) Crash when creating a share https://github.com/owncloud/ios-app/pull/495#issuecomment-548361582  [FIXED]
- [x] (15) Crash creating public links https://github.com/owncloud/ios-app/pull/495#issuecomment-548364233  [FIXED]
- [x] (16) Switching apps makes losing the keyboard https://github.com/owncloud/ios-app/pull/495#issuecomment-548371971 [WONT FIX]

hosy commented on ownCloud/Key Commands/KeyCommands.swift:
```diff
@@ -0,0 +1,1045 @@
+
+	@objc func selectNext(sender: UIKeyCommand) {
+		if let selectedIP = self.tableView?.indexPathForSelectedRow {
+			self.tableView.selectRow(at: NSIndexPath(row: selectedIP.row + 1, section: selectedIP.section) as IndexPath, animated: true, scrollPosition: .middle)
```
Solved by checking bounds

hosy commented on ownCloud/Key Commands/KeyCommands.swift:
```diff
@@ -0,0 +1,1045 @@
+
+	@objc func selectPrevious(sender: UIKeyCommand) {
+		if let selectedIP = self.tableView?.indexPathForSelectedRow {
+			self.tableView.selectRow(at: NSIndexPath(row: selectedIP.row - 1, section: selectedIP.section) as IndexPath, animated: true, scrollPosition: .middle)
```
Solved by checking bounds

hosy commented on ownCloud/Client/Actions/Action.swift:
```diff
@@ -305,6 +309,18 @@ class Action : NSObject {
+		return nil
+	}
+
+	func test() {
```
removed

hosy commented on ownCloud/Client/Actions/Action.swift:
```diff
@@ -305,6 +309,18 @@ class Action : NSObject {
 		})
 	}
 
+	func provideKeyCommand() -> UIKeyCommand? {
```
Cool, that you found this dead code

hosy commented on ownCloud/Key Commands/KeyCommands.swift:
```diff
@@ -0,0 +1,1045 @@
+			})
+			shortcuts.append(favoriteCommand)
+
+			let actionsLocationCollaborate = OCExtensionLocation(ofType: .action, identifier: .collaborateItem)
```
Very good idea!

hosy commented on ownCloud/Key Commands/KeyCommands.swift:
```diff
@@ -0,0 +1,1045 @@
+		let previousObjectCommand = UIKeyCommand(input: UIKeyCommand.inputLeftArrow, modifierFlags: [], action: #selector(selectPrevious), discoverabilityTitle: "Previous".localized)
+
+		var showCommands = false
+		if let firstViewController = self.viewControllers?.first, let pdfController = firstViewController as? PDFViewerViewController {
```
Solved

hosy commented on ownCloud/Key Commands/KeyCommands.swift:
```diff
@@ -0,0 +1,1045 @@
+				if let itemMimeType = item.mimeType {
+					let mimeTypeCF = itemMimeType as CFString
+					if let rawUti = UTTypeCreatePreferredIdentifierForTag(kUTTagClassMIMEType, mimeTypeCF, nil)?.takeRetainedValue() {
+						if core.localCopy(of: item) == nil {
```
Good to know! Thank you!

hosy commented on ownCloud/Key Commands/KeyCommands.swift:
```diff
@@ -0,0 +1,1045 @@
+		}
+	}
+
+	@objc func importPasteboard() {
```
Good idea, moved to an own extension.

hosy commented on ownCloud/Key Commands/KeyCommands.swift:
```diff
@@ -0,0 +1,1045 @@
+				}
+			} else {
+				// System-wide Pasteboard
+				for type in pasteboard.types {
```
Please can you explain?

felix-schwarz replied:
I copied it to the clipboard and then pasted it again. Since the pasteboard returned multiple types here, the file was also pasted multiple times.

hosy replied:
This was moved to a new PR

hosy commented on ownCloud/Key Commands/KeyCommands.swift:
```diff
@@ -0,0 +1,1045 @@
+					let mimeTypeCF = itemMimeType as CFString
+					if let rawUti = UTTypeCreatePreferredIdentifierForTag(kUTTagClassMIMEType, mimeTypeCF, nil)?.takeRetainedValue() {
+						if core.localCopy(of: item) == nil {
+							core.downloadItem(item, options: [ .returnImmediatelyIfOfflineOrUnavailable : true ], resultHandler: { (error, core, item, _) in
```
Good idea. Do we have an HUD class already. Could be needed in general.

hosy replied:
This was moved to a new PR

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/owncloud/ios-app?pullRequest=495) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/owncloud/ios-app?pullRequest=495) before we can accept your contribution.<br/>**3** out of **4** committers have signed the CLA.<br/><br/>:white_check_mark: mneuwert<br/>:white_check_mark: hosy<br/>:white_check_mark: ownclouders<br/>:x: felix-schwarz<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/owncloud/ios-app?pullRequest=495) it.</sub>

hosy commented on ownCloud/Key Commands/KeyCommands.swift:
```diff
@@ -0,0 +1,872 @@
+		if let selectedIndexPath = self.tableView?.indexPathForSelectedRow {
+			let staticRow = staticRowForIndexPath(selectedIndexPath)
+			self.tableView.endEditing(true)
+			if staticRow.type == .switchButton, let switchButon = staticRow.cell?.accessoryView as? UISwitch {
```
That was easy ;)

hosy commented on ownCloud/Client/Actions/Actions+Extensions/UploadMediaAction.swift:
```diff
@@ -25,7 +25,11 @@ class UploadMediaAction: UploadBaseAction {
+	override class var keyCommand : String? { return "M" }
+	override class var keyModifierFlags: UIKeyModifierFlags? { return [.command] }
+
+	private let uploadSerialQueue = DispatchQueue(label: "com.owncloud.upload.queue", target: DispatchQueue.global(qos: .background))
```
removed

hosy commented on ownCloud/Client/Actions/Actions+Extensions/MoveAction.swift:
```diff
@@ -22,7 +22,9 @@ class MoveAction : Action {
-	override class var locations : [OCExtensionLocationIdentifier]? { return [.moreItem, .moreFolder, .toolbar] }
+	override class var locations : [OCExtensionLocationIdentifier]? { return [.moreItem, .moreFolder, .toolbar, .keyboardShortcut] }
+	override class var keyCommand : String? { return "V" }
+	override class var keyModifierFlags: UIKeyModifierFlags? { return [.command] }
```
I changed the keyCommand and keyModifierFlags for 
- CopyAction
- MoveAction
- CutAction
- CopyToPasteboardAction
- ImportPasteboardAction

hosy commented on ownCloud/Client/Actions/Actions+Extensions/MakeAvailableOfflineAction.swift:
```diff
@@ -23,7 +23,9 @@ class MakeAvailableOfflineAction: Action {
-	override class var locations : [OCExtensionLocationIdentifier]? { return [.moreItem, .moreFolder] }
+	override class var locations : [OCExtensionLocationIdentifier]? { return [.moreItem, .moreFolder, .keyboardShortcut] }
+	override class var keyCommand : String? { return "O" }
+	override class var keyModifierFlags: UIKeyModifierFlags? { return [.command] }
```
Changed

hosy commented on ownCloud/Client/Actions/Actions+Extensions/CopyToPasteboardAction.swift:
```diff
@@ -0,0 +1,90 @@
+				return
+			}
+
+			core.downloadItem(item, options: [ .returnImmediatelyIfOfflineOrUnavailable : true ], resultHandler: { (error, core, item, _) in
```
The previous defined `core` not used inside the completion handler, because the completion handler returns it own `core`. So in my opinion `[weak core]` does lead to compile error.

felix-schwarz replied:
Ha, right! Didn't see that. Good catch!

hosy commented on ownCloudTests/Info.plist:
```diff
@@ -17,6 +17,10 @@
+	<string>138</string>
+=======
 	<string>142</string>
+>>>>>>> milestone/1.2
```
fixed

hosy commented on ownCloud/Bookmarks/BookmarkInfoViewController.swift:
```diff
@@ -58,7 +58,7 @@ class BookmarkInfoViewController: StaticTableViewController {
 				self?.present(alertController, animated: true, completion: nil)
 			}
-		}, title: "Include available offline files".localized, value: false, identifier: "row-include-available-offline")
+		}, title: "Include available offline files", value: false, identifier: "row-include-available-offline")
```
fixed

hosy commented on ownCloud/Client/Actions/Actions+Extensions/ImportPasteboardAction.swift:
```diff
@@ -0,0 +1,120 @@
+					}
+				} else if let data = pasteboard.data(forPasteboardType: "com.owncloud.uti.OCItem.cut"), let object = NSKeyedUnarchiver.unarchiveObject(with: data) {
+					if let item = object as? OCItem, let name = item.name {
+						core.copy(item, to: rootItem, withName: name, options: nil, resultHandler: { (error, _, _, _) in
```
fixed

hosy commented on ownCloud/Client/Actions/Actions+Extensions/ImportPasteboardAction.swift:
```diff
@@ -0,0 +1,120 @@
+						core.copy(item, to: rootItem, withName: name, options: nil, resultHandler: { (error, _, _, _) in
+							if error != nil {
+							} else {
+								core.delete(item, requireMatch: true) { (_, _, _, _) in
```
no longer needed, because `copy` and `delete` was removed by `move`

hosy commented on ownCloud/Key Commands/KeyCommands.swift:
```diff
@@ -0,0 +1,872 @@
+extension BookmarkViewController {
+	override var keyCommands: [UIKeyCommand]? {
+		return [
+			UIKeyCommand(input: UIKeyCommand.inputEscape, modifierFlags: [], action: #selector(dismiss), discoverabilityTitle: "Cancel".localized)
```
Pressing `Control` and `[` will achieve the ESC escape key function. Apple also use the ESC key for many keyboard commands and I think we should follow this way.

felix-schwarz replied:
TIL! Thanks!

hosy commented on ownCloud/Key Commands/KeyCommands.swift:
```diff
@@ -0,0 +1,872 @@
+
+		var shortcuts = [UIKeyCommand]()
+		if let leftItem = self.navigationItem.leftBarButtonItem, let action = leftItem.action {
+			let dismissCommand = UIKeyCommand(input: UIKeyCommand.inputEscape, modifierFlags: [], action: action, discoverabilityTitle: "Cancel".localized)
```
Pressing `Control` and `[` will achieve the ESC escape key function. Apple also use the ESC key for many keyboard commands and I think we should follow this way.

hosy commented on ownCloud/Key Commands/KeyCommands.swift:
```diff
@@ -0,0 +1,872 @@
+			shortcuts.append(contentsOf: superKeyCommands)
+		}
+		let searchCommand = UIKeyCommand(input: "F", modifierFlags: [.command], action: #selector(enableSearch), discoverabilityTitle: "Search".localized)
+		let doneCommand = UIKeyCommand(input: UIKeyCommand.inputEscape, modifierFlags: [], action: #selector(dismiss), discoverabilityTitle: "Done".localized)
```
Pressing `Control` and `[` will achieve the ESC escape key function. Apple also use the ESC key for many keyboard commands and I think we should follow this way.

hosy commented on ownCloud/Key Commands/KeyCommands.swift:
```diff
@@ -0,0 +1,872 @@
+		if let superKeyCommands = super.keyCommands {
+			shortcuts.append(contentsOf: superKeyCommands)
+		}
+		let dismissCommand = UIKeyCommand(input: UIKeyCommand.inputEscape, modifierFlags: [], action: #selector(dismiss), discoverabilityTitle: "Cancel".localized)
```
Pressing `Control` and `[` will achieve the ESC escape key function. Apple also use the ESC key for many keyboard commands and I think we should follow this way.

hosy commented on ownCloud/Key Commands/KeyCommands.swift:
```diff
@@ -0,0 +1,872 @@
+		if let superKeyCommands = super.keyCommands {
+			shortcuts.append(contentsOf: superKeyCommands)
+		}
+		let doneCommand = UIKeyCommand(input: UIKeyCommand.inputEscape, modifierFlags: [], action: #selector(dismiss), discoverabilityTitle: "Done".localized)
```
Pressing `Control` and `[` will achieve the ESC escape key function. Apple also use the ESC key for many keyboard commands and I think we should follow this way.

hosy commented on ownCloud/Key Commands/KeyCommands.swift:
```diff
@@ -0,0 +1,872 @@
+		if let superKeyCommands = super.keyCommands {
+			shortcuts.append(contentsOf: superKeyCommands)
+		}
+		let dismissCommand = UIKeyCommand(input: UIKeyCommand.inputEscape, modifierFlags: [], action: #selector(dismiss), discoverabilityTitle: "Cancel".localized)
```
Pressing `Control` and `[` will achieve the ESC escape key function. Apple also use the ESC key for many keyboard commands and I think we should follow this way.

hosy commented on ownCloud/Key Commands/KeyCommands.swift:
```diff
@@ -0,0 +1,872 @@
+
+		let createFolder = UIKeyCommand(input: "N", modifierFlags: [.command], action: #selector(createFolderButtonPressed), discoverabilityTitle: "Create Folder".localized)
+		shortcuts.append(createFolder)
+		let dismissCommand = UIKeyCommand(input: UIKeyCommand.inputEscape, modifierFlags: [], action: #selector(dismiss), discoverabilityTitle: "Cancel".localized)
```
Pressing `Control` and `[` will achieve the ESC escape key function. Apple also use the ESC key for many keyboard commands and I think we should follow this way.

hosy commented on ownCloud/Key Commands/KeyCommands.swift:
```diff
@@ -0,0 +1,872 @@
+        if cancelButton?.isHidden == false {
+            keyCommands.append(
+
+                UIKeyCommand(input: UIKeyCommand.inputEscape,
```
Pressing `Control` and `[` will achieve the ESC escape key function. Apple also use the ESC key for many keyboard commands and I think we should follow this way.

hosy commented on ownCloud/Key Commands/KeyCommands.swift:
```diff
@@ -0,0 +1,872 @@
+		let selectAllCommand = UIKeyCommand(input: "A", modifierFlags: [.command], action: #selector(selectAllItems), discoverabilityTitle: "Select All".localized)
+		let deselectAllCommand = UIKeyCommand(input: "D", modifierFlags: [.command], action: #selector(deselectAllItems), discoverabilityTitle: "Deselect All".localized)
+		let uploadCommand = UIKeyCommand(input: "U", modifierFlags: [.command], action: #selector(upload), discoverabilityTitle: "Upload".localized)
+		let dismissCommand = UIKeyCommand(input: UIKeyCommand.inputEscape, modifierFlags: [], action: #selector(dismiss), discoverabilityTitle: "Cancel".localized)
```
Pressing `Control` and `[` will achieve the ESC escape key function. Apple also use the ESC key for many keyboard commands and I think we should follow this way.

hosy commented on ownCloud/Key Commands/KeyCommands.swift:
```diff
@@ -0,0 +1,872 @@
+			shortcuts.append(nextObjectCommand)
+		}
+
+		let dismissCommand = UIKeyCommand(input: UIKeyCommand.inputEscape, modifierFlags: [], action: #selector(dismiss), discoverabilityTitle: "Cancel".localized)
```
Pressing `Control` and `[` will achieve the ESC escape key function. Apple also use the ESC key for many keyboard commands and I think we should follow this way.

hosy commented on ownCloud/Key Commands/KeyCommands.swift:
```diff
@@ -0,0 +1,872 @@
+        switch key {
+        case "\u{8}":
+            deleteLastDigit()
+        case UIKeyCommand.inputEscape:
```
Pressing `Control` and `[` will achieve the ESC escape key function. Apple also use the ESC key for many keyboard commands and I think we should follow this way.

jesmrec commented:
### (1)

1. Select an item in the list and Move (cmd + V)
2. In the picker, select a correct target to move by browsing

Current: No way to submit the selection. By long pressing the cmd key (to get the available commands), the options displayed are:

![Screenshot 2019-10-23 at 16 23 54](https://user-images.githubusercontent.com/14894746/67403105-c8e55500-f5b1-11e9-94d6-e3103ca2b8cc.png)

Move is twice:

1. cmd + V -> to open the picker
2. alt + cmd + v -> it opens the picker as well

Expected:

Key shortcut to move.

NOTE: same behaviour for copy

IPadPro iOS13.1.3
ownCloud 10.2.1

jesmrec commented:
### (2)

1. Browse over an item and select Create Link (cmd + L)
2. Select Create Public Link
3. Try to modify the expiration date

Current: i did not find the way to do it

jesmrec commented:
### (3)

About the "Search" (cmd +F if no items are selected). In case any item in the list is selected, the only way to search is browsing back and then, browsing forward again to deselect. Not a bad behaviour, but not intuitive. What do you think?

jesmrec commented:
### (4)

There is not way to create a folder (cmd + N) in the root folder. Option only available by selecting  a folder in the list. Then, the folder is created inside. 

Is this the expected behaviour?

jesmrec commented:
### (5)

1. Select an item in the list
2. Make it as available offline (alt + cmd + O)

Current: all files in the list are set as av. offline
Expected: Only the selected file is set as av. offline

IPadPro iOS13.1.3
ownCloud 10.2.1

jesmrec commented:
### (6)

I miss a couple of options:

- Option to unset as available offline
- Option to unfavorite

jesmrec commented:
### (7)

Regarding authentication: 

- If any issue happen in the URL validation, there is no command to approve it
- If the authentication is OAuth2, after typing the URL, no command to continue.

jesmrec commented:
### (8)

1. Select an account from the list
2. manage account (caps + cmd + M)

Current: No way to cancel the view

IPadPro iOS13.1.3
ownCloud 10.2.1
app commit 0156427

jesmrec commented:
### (9)

In Settings view:

Delete unused copies: No way to change the value in the slider

hosy commented:
> ### (2)
> 1. Browse over an item and select Create Link (cmd + L)
> 2. Select Create Public Link
> 3. Try to modify the expiration date
> 
> Current: i did not find the way to do it
> 
> IPadPro iOS13.1.3
> ownCloud 10.2.1
> app commit `0156427`

Can you create an issue for this? Do not find a way to do this in an easy way.

hosy commented:
@jesmrec 
Fixed a couple of findings:
- fixed QA finding (1), directory picker key commands fixed 
- fixed QA finding (4), create folder in root folder
- fixed QA finding (5), available offline for selected item 
- fixed QA finding (6), make unavailable, unfavorite
- fixed QA finding (8), dismiss manage account view

jesmrec commented:
### (10)

Detected while checking the fix for (1)

1. Select an item in the list and Move (alt + cmd + V)
2. In the picker move down through the items with the down arrow

Current: When you reach the end of the list in the picker, the app crashes
Expected: No crash

IPadPro iOS13.1.3
ownCloud 10.2.1
app commit `c0c81024`

jesmrec commented:
(1), (4), (8) -> fixed

(2) -> will create new issue

(5) -> now, i do not see any option to make stuff as av. offline. The former command (alt + cmd + O) is now used to open in a new window

(6) -> unfavorite is OK. But no option to make unavailable offline.

(3), (7), (9) -> Pending

besides, i reported (10)

hosy commented:
@jesmrec 
please check (3), (5), (10)

hosy commented:
@jesmrec 
fixed findings (3), (7), (9), (11)
- Cancel Search
- Commands for internal Alert Views (OAuth alert and web views cannot be controlled via keyboard commands, because this came from other app process)
- Change slider value
- Crash in welcome view

jesmrec commented:
> fixed findings (3), (7), (9), (11)

everything fixed.

I will perform a 2nd QA round.

jesmrec commented:
### (12)

The option to create account from the welcome view (cmd + "+") is lost. It appears in the command list, but it does not work

jesmrec commented:
### (13)

1. Select an item in the list and the commando to delete (cmd + back arrow) -> Dialog to delete is displayed

Long press on cmd to check the options available:

Option to delete: cmd +1 -> works fine
Option to cancel: cmd + 2 -> does not work
Also, appears a 3rd option to switch to Status view. This one is not necessary here

IPadPro iOS13.1.3
ownCloud 10.3
app commit `71a90c9`

hosy commented:
@jesmrec 
fixed finding (13)
- Cancel did not dismiss alert view
- Only show alert view key commands

(12) is working after the second try, I think this is a iOS related bug and I would suggest: (Won't fix)

jesmrec commented:
### (14)

1. Select an item in the file list
2. cmd + S (sharing)
3. cmd + F (search)
4. type the user name to share with and press "Enter"

Current: You can move up and down to select a user from the matches list. If you press right arrow, app crashes
Expected: With the right arrow, the dialog to create the share is displayed.

IPadPro iOS13.1.3
ownCloud 10.3
app commit `fbe97e8f`

jesmrec commented:
### (15)

1. Select an item from the list
2. cmd + L (create public link)

Current: crash
Expected: creation of link

IPadPro iOS13.1.3
ownCloud 10.3
app commit `fbe97e8f`

jesmrec commented:
### (16)

Maybe not an app issue, just for clarification:

In you switch from oC to other app, and after a while, you switch again to oC, the keyboard control is lost (nothing works).You have to kill the app and open it again in order the keyboard to work again.

hosy commented:
@jesmrec (14), (15) is fixed, (16) is an Apple iOS bug

jesmrec commented:
With the fixed (14) and (15) i can check the sharing options. Two things (probably the last ones):

(17): create a public link. Then, open the link and edit the name of the link -> app crashes

I miss an command to share the public link.

hosy commented:
@jesmrec finding (17) fixed, sharing command is now available, sharing view cannot be controlled with keyboard commands from the app

jesmrec commented:
(17) is fixed and the new command is welcome.

jesmrec commented:
Feature approved
